### PR TITLE
Add flag to enable SwitchDev mode in SriovNetworkNodePolicy CRD

### PR DESCRIPTION
**Pull request by pliurh**
The generic plugin will skip configuring swichdev devices. It shall be done
through a systemd services on the host.

### CODE REVIEW

**Comment by pliurh** [comment]
@moshe010 @Mmduh-483 @adrianchiris PTAL.

File: pkg/utils/utils.go [hunk 1]
```diff
@@ -115,6 +118,10 @@ func SyncNodeState(newState *sriovnetworkv1.SriovNetworkNodeState) error {
 		for _, iface := range newState.Spec.Interfaces {
 			if iface.PciAddress == ifaceStatus.PciAddress {
 				configured = true
+				if iface.SwitchMode == sriovnetworkv1.SWITCHMODE_SWITCHDEV {
+					// Skip for switchdev devices
+					continue
```
Comment (zshi-redhat) [inline pkg/utils/utils.go:123]: nit: break

File: pkg/plugins/generic/generic_plugin.go [hunk 2]
```diff
@@ -186,6 +186,10 @@ func needDrainNode(desired sriovnetworkv1.Interfaces, current sriovnetworkv1.Int
 		for _, iface := range desired {
 			if iface.PciAddress == ifaceStatus.PciAddress {
 				configured = true
+				if iface.SwitchMode == sriovnetworkv1.DEVMODE_SWITCHDEV {
+					// Skip for switchdev devices
+					continue
```
Comment (zshi-redhat) [inline pkg/plugins/generic/generic_plugin.go:191]: nit: break

File: pkg/utils/utils.go [hunk 3]
```diff
@@ -115,6 +118,10 @@ func SyncNodeState(newState *sriovnetworkv1.SriovNetworkNodeState) error {
 		for _, iface := range newState.Spec.Interfaces {
 			if iface.PciAddress == ifaceStatus.PciAddress {
 				configured = true
+				if iface.SwitchMode == sriovnetworkv1.SWITCHMODE_SWITCHDEV {
+					// Skip for switchdev devices
```
Comment (zshi-redhat) [inline pkg/utils/utils.go:122]: Add a debug log here for skipping switchdev mode device.

File: pkg/plugins/generic/generic_plugin.go (hunk 2 above)
Comment (zshi-redhat) [inline pkg/plugins/generic/generic_plugin.go:191]: Add a debug log.

File: api/v1/helper.go [hunk 4]
```diff
@@ -22,6 +22,8 @@ const (
 	MANIFESTS_PATH       = "./bindata/manifests/cni-config"
 	LASTNETWORKNAMESPACE = "operator.sriovnetwork.openshift.io/last-network-namespace"
 	FINALIZERNAME        = "netattdef.finalizers.sriovnetwork.openshift.io"
+	SWITCHMODE_LEGACY    = "legacy"
```
Comment (zshi-redhat) [inline api/v1/helper.go:25]: SWITCHMODE - > ESWITCHMODE ?

File: api/v1/sriovnetworknodepolicy_types.go [hunk 5]
```diff
@@ -35,6 +35,9 @@ type SriovNetworkNodePolicySpec struct {
 	// +kubebuilder:validation:Enum=eth;ETH;ib;IB
 	// NIC Link Type. Allowed value "eth", "ETH", "ib", and "IB".
 	LinkType string `json:"linkType,omitempty"`
+	// +kubebuilder:validation:Enum=legacy;switchdev
+	// NIC Device Mode. Allowed value "legacy","switchdev".
+	SwitchMode string `json:"switchMode,omitempty"`
```
Comment (zshi-redhat) [inline api/v1/sriovnetworknodepolicy_types.go:40]: SwitchMode -> EswitchMode ?

File: api/v1/sriovnetworknodestate_types.go [hunk 6]
```diff
@@ -20,6 +20,7 @@ type Interface struct {
 	Mtu        int       `json:"mtu,omitempty"`
 	Name       string    `json:"name,omitempty"`
 	LinkType   string    `json:"linkType,omitempty"`
+	SwitchMode string    `json:"switchMode,omitempty"`
```
Comment (zshi-redhat) [inline api/v1/sriovnetworknodestate_types.go:23]: EswitchMode?

File: api/v1/sriovnetworknodestate_types.go [hunk 7]
```diff
@@ -43,6 +44,7 @@ type InterfaceExt struct {
 	NumVfs     int               `json:"numVfs,omitempty"`
 	LinkSpeed  string            `json:"linkSpeed,omitempty"`
 	LinkType   string            `json:"linkType,omitempty"`
+	SwitchMode string            `json:"switchMode,omitempty"`
```
Comment (zshi-redhat) [inline api/v1/sriovnetworknodestate_types.go:47]: EswitchMode

**Review by zshi-redhat — COMMENTED** [review]
How do we handle nodeState syncStatus for switchdev policy?
Until MCO reboots the node, will syncStatus be succeeded?

**Review by pliurh — COMMENTED** [review]
> How do we handle nodeState syncStatus for switchdev policy?
> Until MCO reboots the node, will syncStatus be succeeded?

That's a good question. In this patch, the config daemon doesn't handle the syncStatus for switchdev policy. It will just become 'Succeed'. If we make it return a failure, the config daemon would keep retrying. I am wondering whether we have a better way to handle it.

**Comment by zshi-redhat** [comment]
> > How do we handle nodeState syncStatus for switchdev policy?
> > Until MCO reboots the node, will syncStatus be succeeded?
> 
> That's a good question. In this patch, the config daemon doesn't handle the syncStatus for switchdev policy. It will just become 'Succeed'. If we make it return a failure, the config daemon would keep retrying. I am wondering whether we have a better way to handle it.

If cluster nodes are applied with mixed policies (switchdev and non-switchdev), would there be a problem that MCO and sriov config daemon are not synced on which node is being rebooted, which may result in interrupted workloads?

Maybe we need a sync mechenism between Operator applying MachineConfig and config daemon?
Shall we use sriov config daemon to write the device switchdev systemd config?

File: config/crd/bases/sriovnetwork.openshift.io_sriovibnetworks.yaml [hunk 8]
```diff
@@ -1,6 +1,6 @@
 
 ---
-apiVersion: apiextensions.k8s.io/v1beta1
+apiVersion: apiextensions.k8s.io/v1
```
Comment (Mmduh-483) [inline config/crd/bases/sriovnetwork.openshift.io_sriovibnetworks.yaml:3]: I think this is not related to the switchdev

Comment (Mmduh-483) [reply config/crd/bases/sriovnetwork.openshift.io_sriovibnetworks.yaml:3]: There are several CRD files changes not related to eswitch mode, it seems you are not using the same version of the operator-sdk, IMO it can be in another PR

File: pkg/plugins/generic/generic_plugin.go [hunk 9]
```diff
@@ -186,6 +186,11 @@ func needDrainNode(desired sriovnetworkv1.Interfaces, current sriovnetworkv1.Int
 		for _, iface := range desired {
 			if iface.PciAddress == ifaceStatus.PciAddress {
```
Comment (Mmduh-483) [inline pkg/plugins/generic/generic_plugin.go:187]: No need to continue the loop after finding the device, maybe adding break at end of if

File: pkg/utils/utils.go [hunk 10]
```diff
@@ -595,3 +603,12 @@ func generateRandomGuid() net.HardwareAddr {
 
 	return guid
 }
+
+func GetNicSriovMode(pciAddress string) (string, error) {
```
Comment (Mmduh-483) [inline pkg/utils/utils.go:607]: @pliurh Will `SetNicSriovMode` be in a follow up?

File: Makefile [hunk 11]
```diff
@@ -39,7 +39,7 @@ BUNDLE_METADATA_OPTS ?= $(BUNDLE_CHANNELS) $(BUNDLE_DEFAULT_CHANNEL)
 # Image URL to use all building/pushing image targets
 IMG ?= controller:latest
 # Produce CRDs that work back to Kubernetes 1.11 (no version conversion)
-CRD_OPTIONS ?= "crd:trivialVersions=true"
+CRD_OPTIONS ?= "crd:crdVersions={v1},trivialVersions=true"
```
Comment (adrianchiris) [inline Makefile:42]: can the CRD version bump be in a separate commit in this PR ? or a separate PR ?

i.e recreate CRDs with the new operator SDK without switchdev change

File: api/v1/helper.go [hunk 12]
```diff
@@ -19,9 +19,11 @@ import (
 )
 
 const (
-	MANIFESTS_PATH       = "./bindata/manifests/cni-config"
-	LASTNETWORKNAMESPACE = "operator.sriovnetwork.openshift.io/last-network-namespace"
-	FINALIZERNAME        = "netattdef.finalizers.sriovnetwork.openshift.io"
+	MANIFESTS_PATH        = "./bindata/manifests/cni-config"
+	LASTNETWORKNAMESPACE  = "operator.sriovnetwork.openshift.io/last-network-namespace"
+	FINALIZERNAME         = "netattdef.finalizers.sriovnetwork.openshift.io"
+	ESWITCHMODE_LEGACY    = "legacy"
```
Comment (adrianchiris) [inline api/v1/helper.go:25]: note to self These should really be constants in netlink package.

**Review by adrianchiris — COMMENTED** [review]
So i would be happy if the commit message provided more context on:

1. The problem this PR is trying to solve
2. How it attempts to solve it (general outline of the changes)

Also an outline on how this will work in the full solution is appreciated.

**Comment by adrianchiris** [comment]
Adding this for reference. https://github.com/openshift/enhancements/pull/296

**Comment by adrianchiris** [comment]
> Also an outline on how this will work in the full solution is appreciated.
as well as more information on the systemd service, what it will do and who will deploy it.

File: config/crd/bases/sriovnetwork.openshift.io_sriovibnetworks.yaml (hunk 8 above)
Comment (pliurh) [reply config/crd/bases/sriovnetwork.openshift.io_sriovibnetworks.yaml:3]: Yeah, it can be in a separate PR. Those changes were mistakenly introduced by the previous operator upgrade.

File: pkg/utils/utils.go (hunk 10 above)
Comment (pliurh) [reply pkg/utils/utils.go:607]: No, at least not for openshift. The plan is to use the Machine Config Operator to insert a new systemd service which will configure the VF and set NIC to switchdev mode on the host.

Comment (pliurh) [reply pkg/utils/utils.go:607]: In this release, we only use switchdev for OVN offloading use case. And we want users can use VF bond with it. In openshift-sdn, bond is created by NetworkManager configure file inserted by MCO. So we have to configure switchdev before NetworkManager starts on the host.

File: pkg/plugins/generic/generic_plugin.go (hunk 9 above)
Comment (pliurh) [reply pkg/plugins/generic/generic_plugin.go:187]: fixed

File: Makefile (hunk 11 above)
Comment (pliurh) [reply Makefile]: Addressed by https://github.com/openshift/sriov-network-operator/pull/389

**Comment by pliurh** [comment]
> as well as more information on the systemd service, what it will do and who will deploy it.

PTAL https://github.com/openshift/sriov-network-operator/pull/393

**Comment by pliurh** [comment]
@zshi PTAL

File: pkg/utils/utils.go [hunk 13]
```diff
@@ -115,6 +118,11 @@ func SyncNodeState(newState *sriovnetworkv1.SriovNetworkNodeState) error {
 		for _, iface := range newState.Spec.Interfaces {
 			if iface.PciAddress == ifaceStatus.PciAddress {
 				configured = true
+				if iface.EswitchMode == sriovnetworkv1.ESWITCHMODE_SWITCHDEV {
```
Comment (zshi-redhat) [inline pkg/utils/utils.go:121]: @pliurh  this part will be updated in the following PR, right? e.g. To not skip switchdev device in daemon.

Comment (pliurh) [reply pkg/utils/utils.go]: removed.
